---
title: "PedFac performance on empricial pedigrees - ch2"
author: "Thomas C Ng"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: yes
  html_notebook:
    toc: yes
---

```{r setup, include=FALSE}
library(knitr)
library(tidyverse)
library(pedfac)
library(tictoc) #devtools::install_github("jabiru/tictoc")
library(sequoia)

opts_chunk$set(#results = 'asis',      # This is essential (can also be set at the chunk-level)
                #comment = NA, 
                #prompt = FALSE, 
                cache = FALSE)
```


We will start out simple by revisiting the two generation full-sib pedigree from Anderson and Ng to make sure that we are on the right track. This simulated genotype data PedFac's acyclic sampler is generated from an empirical two generation chinook pedigree with the absence of the parental generation. We will simulate six replicate cases of 95 SNPs with allele frequency of ~\beta{0.5,0.5} with 0.5% rate in missing genotype and 2% rate in genotype error. 

We report full sib assignment ROC ()

```{r}
#retrieve from https://github.com/eriqande/tpb-colony-compare/tree/master/data
load("/Users/thomasn/repo/pedigree-factor-graphs/data/andersonng/chinook_full_sibs.Rda")

#double check that it's all full sib fam
chinook_full_sibs_pedigree %>%
  group_by(Pa) %>%
  summarise(num.Ma = length(unique(Ma))) %>%
  ungroup() %>%
  filter(num.Ma > 1) %>%
  nrow()

chinook_full_sibs_pedigree %>%
  group_by(Pa) %>%
  summarise(num.kids = length(unique(Kid))) %>%
  ungroup() %>% select(num.kids) %>% table
```
```{r}
chinook_full_sibs_pedigree %>%
  group_by(Pa, Ma) %>%
  summarise(n=n()) %>%
  ungroup() %>%
  summarise(mean.size = mean(n))
```

```{r}
## simulate the two generation full sib data and run pedigraph with it

simFulldata <- function(seed.num = 234,
                        pedigraph.version = 0.185,
                        pedigraph.path = "/Users/thomasn/repo/pedigree-factor-graphs/src",
                        outdir.path="/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib", n.iter = 10,
                        replicate.indx  = 1,
                        sf.rate = 1,
                        ped.df = chinook_full_sibs_pedigree,
                        n.snp =95, geno.err = 0.02, missing.unk = 0.005,
                        pedfac.opt = "",
                        rerun= TRUE,
                        pastLS = NA,
                        run.label = "pedfac",
                        case.label=paste0("sf_",sf.rate)) {
  
  if (!rerun && is.na(pastLS) ) {stop("need past LS")}
  run.summary <- list()
  
  run.summary$param <- tibble::lst(seed.num, pedigraph.version, pedigraph.path, outdir.path, replicate.indx, sf.rate, n.snp, geno.err, missing.unk, pedfac.opt)
  
  ## factorize ped.df
  mating.factor <- factor(unlist(ped.df))
  mating.lvl <- levels(mating.factor)
  mating.factor.tbl <- matrix(as.numeric(mating.factor), ncol=3) %>% as.tibble()
  colnames(mating.factor.tbl) <- c("kid","pa","ma")
  
  outfolder.path <- paste0(outdir.path,"/",case.label,"/",replicate.indx)
  
  pedfac.out.path <- paste0(outdir.path,"/",case.label,"/",replicate.indx,"/",run.label)
  
  if (rerun) {
  dir.create(file.path(outfolder.path), recursive = TRUE)
  dir.create(file.path(pedfac.out.path), recursive = TRUE)
  }
  
  set.seed(seed.num)
  random.arr <- ceiling(c(runif(2,0,1e6)))
  
  
  if (rerun) {
    write.table(random.arr, paste0(outfolder.path,"/","rand"),
              sep = " ",eol = " ",quote = FALSE, col.names = FALSE, row.names = FALSE)
    
    geno.input.tbl <- simGeno(mating.tbl = mating.factor.tbl, out.path = outfolder.path,
          n.snp =n.snp, geno.err = geno.err, random.seed = seed.num,sampling.frac.rate = sf.rate,sample.frac.gen = 1, missing.unk = missing.unk) 
  
  } else {
    geno.input.tbl <- pastLS$geno.input.tbl
  }
  
  run.summary$geno.input.tbl <- geno.input.tbl
  
  
  ## assume indiv ID is preserves
  
  
  pedfac.call <- paste0(pedigraph.path,"/pedigraph_",pedigraph.version, " -d ", outfolder.path,
                         " -n ", n.iter, 
                         " -r ",outfolder.path,"/rand ", pedfac.opt, collapse = "")
  
  if (rerun) {
  tic("run pedFac:")
  system(pedfac.call,ignore.stderr = F, ignore.stdout = F)
  runtime <- toc()
  write.table(runtime$toc - runtime$tic, file =paste0(outfolder.path,"/",run.label,"/pedFac.time"), quote = FALSE, row.names = FALSE, col.names = FALSE)
  run.summary$runtime <- runtime$toc - runtime$tic
  } else {
    #runtime <- read.table(file = paste0(outfolder.path,"/",run.label,"/pedFac.time"),stringsAsFactors = FALSE)  %>% as.numeric()
    run.summary$runtime <- pastLS$runtime
  }
  
  if (rerun) {
  pedfac.outfile.ls <- list.files(paste0(outfolder.path,"/out"), include.dirs=TRUE, full.names = T)
  file.copy(pedfac.outfile.ls,
            paste0(pedfac.out.path), overwrite = T)
  }
  
  run.summary$pedfac.cmd <- pedfac.call
  
  out.ped.raw.tbl <- read.table(paste0(outfolder.path, "/out/ped.txt",collapse = ""), stringsAsFactors = FALSE) %>%
    dplyr::tbl_df() %>%
    dplyr::rename("iter"="V1", "kid"="V2", "pa"="V3", "ma"="V4") 
  
  id.ls <-readRDS(paste0(outfolder.path, "/id.rds",collapse = ""))$id
  max.id <- max(id.ls)
  
  out.ll.tbl <- read.table(paste0(outfolder.path, "/out/ll.txt",collapse = ""), stringsAsFactors = FALSE) %>%
    dplyr::tbl_df() %>%
    dplyr::rename("iter"="V1", "kid"="V2", "ll"="V3")
  
  run.summary$out.ll.tbl <- out.ll.tbl
  
  # max.ll.indx <- which.max(out.ll.tbl$ll)
  # max.ll.sampled.iter <- out.ll.tbl$iter[max.ll.indx]
  # max.ll.sampled.id <- out.ll.tbl$kid[max.ll.indx]
  # 
  # max.ll.ped.indx <-  intersect(which(out.ped.raw.tbl$iter == max.ll.sampled.iter),
  #                            which(out.ped.raw.tbl$kid == max.ll.sampled.id))
  
  # if (max.ll.sampled.iter == 0) {
  #   maxLL.ped <- out.ped.raw.tbl %>% filter(iter == max.ll.sampled.iter) %>%
  #     mutate(pa = ifelse(row_number()>max.ll.ped.indx, -1, pa ),
  #            ma = ifelse(row_number()>max.ll.ped.indx, -1, ma )) %>%
  #     ungroup() %>%
  #     select(-iter)
  # } else {
  #   maxLL.ped <- out.ped.raw.tbl[1:max.ll.ped.indx,] %>%
  #     filter(iter >= max.ll.sampled.iter-1) %>%
  #     arrange(desc(iter)) %>%
  #     group_by(kid) %>%
  #     summarise(pa = pa[1],
  #               ma = ma[1])
  # }
  
  
  # maxLL.ped.tbl <- maxLL.ped %>% group_by(kid, pa, ma) %>%
  #   mutate(kid.id =  ifelse(kid >=0, subbingID(kid, max.id, id.ls), -1), 
  #          pa.id = ifelse(pa >=0,subbingID(pa, max.id, id.ls), -1),
  #          ma.id = ifelse(ma >=0, subbingID(ma, max.id, id.ls), -1))
  
  out.ped.tbl <- out.ped.raw.tbl %>%
    mutate(order = row_number())%>%
    arrange(iter, desc(order))%>%
    group_by(iter, kid)%>%
    summarise(pa = pa[1], ma =ma[1])
  
  run.summary$out.ped.tbl <- out.ped.tbl
  
  if (sf.rate != 0) {
    message("working out parentage inference:")
    parentage.grp.infer <- retrieveParent(out.ped.tbl, max.id, id.ls)
    
    observed.id.ls <- geno.input.tbl$id
    parentage.grp.truth <- mating.factor.tbl %>% ungroup() %>%
      summarise(kid.id = ifelse(kid %in% observed.id.ls, kid, -1),
             pa.T = ifelse(pa %in% observed.id.ls, pa, -1),
             ma.T = ifelse(ma %in% observed.id.ls, ma, -1))  %>%
      filter(kid.id != -1)
    
    run.summary$parent.prob.tbl <- left_join(parentage.grp.infer, parentage.grp.truth, by=c("kid.id")) %>%
      ungroup() %>%
      mutate(n.match = (pa.id == pa.T)+(ma.id == ma.T))
    
    # run.summary$parent.maxll.tbl <- left_join(maxLL.ped.tbl %>% ungroup() %>% filter(kid.id != -1) %>% select(kid.id, pa.id, ma.id), parentage.grp.truth, by=c("kid.id")) %>%
    #   ungroup() %>%
    #   mutate(n.match = (pa.id == pa.T)+(ma.id == ma.T))
    
    
  }
  
  message("writing fullsib_assignment.txt ")
  full.sib.grp.infer <- RetrieveFullSib(out.ped.tbl, max.id, id.ls)
  
  write.table(full.sib.grp.infer,
              paste0(pedfac.out.path,"/fullsib_assignment.txt"),
              sep = " ",eol = "\n",quote = FALSE, col.names = FALSE, row.names = FALSE)
  
  #full.sib.pair.maxll.tbl <- RetrieveFullSibTruth(maxLL.ped.tbl %>% ungroup() %>% filter(kid.id != -1) %>% select(kid, pa, ma)) %>% rename(presence.maxll = "presence.T")
  
  ## perform pairwise sib-pair from the truth ped
  full.sib.pair.truth.tbl <- RetrieveFullSibTruth(mating.factor.tbl)
  ## run cluster analysis
  
  run.summary$pairwise.tbl <- full_join(full.sib.grp.infer, full.sib.pair.truth.tbl, by= c("kid.1", "kid.2"))
  #run.summary$fullsib.pairwise.maxll <- full_join(full.sib.pair.maxll.tbl, full.sib.pair.truth.tbl, by= c("kid.1", "kid.2"))
  run.summary$fullsib.cluster.ls <- RetrieveFullSibGrp(out.ped.tbl, mating.factor.tbl, max.id, id.ls)
  #run.summary$maxLL.fullsib.cluster.ls <- RetrieveFullSibGrp(maxLL.ped.tbl, mating.factor.tbl, max.id, id.ls,is.single.ped.entry = T)
  
  return(run.summary)
}
```

```{r}
RunSequoia <- function(seed.num = 234,
                        outdir.path="/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib", 
                        replicate.indx  = 1,
                        sf.rate = 1,
                        ped.df = chinook_full_sibs_pedigree,
                        n.snp =95, geno.err = 0.02, missing.unk = 0.005,
                       rerun =TRUE,
                      pastLS = NA,
                      case.label=paste0("sf_",sf.rate)) {
  
  if (!rerun && is.na(pastLS) ) {stop("need past LS")}
  
  run.summary <- list()
  
  run.summary$param <- tibble::lst(seed.num, outdir.path, replicate.indx, sf.rate, n.snp, geno.err, missing.unk)
  
  ## require that pedFac has already run Sim
  ## factorize ped.df
  mating.factor <- factor(unlist(ped.df))
  mating.lvl <- levels(mating.factor)
  mating.factor.tbl <- matrix(as.numeric(mating.factor), ncol=3) %>% as.tibble()
  colnames(mating.factor.tbl) <- c("kid","pa","ma")
  
  outfolder.path <- paste0(outdir.path,"/",case.label,"/",replicate.indx)
  pedFac.geno.tbl <- read.table(paste0(outfolder.path,"/","geno.txt"), stringsAsFactors = F) %>%
    as_tibble()
  
  lifeHistory.tbl <- pedFac.geno.tbl %>%
    group_by(V1) %>%
    summarise(ID = V1,
           Sex = as.integer(3-V3),
           BirthYear = V4) %>%
    ungroup() %>%
    select(ID, Sex, BirthYear) %>% data.frame()
  
  geno.matrix <- pedFac.geno.tbl %>% ungroup() %>% select(-V1, -V2, -V3, -V4, -V5) %>%
    as.matrix
  
  geno.matrix[geno.matrix == 3] <- -9
  rownames(geno.matrix) <- lifeHistory.tbl$ID
  
  set.seed(seed.num)
  
  if (rerun) {
  tic("run sequoia:")
  
  err.rate <- geno.err
  error.rate.0 <- 1-(err.rate)-(err.rate/2)^2
  error.rate.1 <- (err.rate/2)^2
  error.rate.2 <- err.rate /2
  error.rate.3 <- 1-err.rate

  #ErrM <<- matrix(c(error.rate.0,err.rate,error.rate.1,
   #       error.rate.2, 1-err.rate, error.rate.2,
    #      error.rate.1,err.rate,error.rate.0), ncol = 3)
  
  ErrM <<- matrix(c(error.rate.3,error.rate.2,error.rate.2,
                    error.rate.2, error.rate.3,error.rate.2,
                    error.rate.2,error.rate.2,error.rate.3), ncol = 3)
  
  if (sf.rate != 0) {      
  ParOUT_Mono <- sequoia(GenoM = as.matrix(geno.matrix), LifeHistData = lifeHistory.tbl, MaxSibIter = 0,Complex = "mono", Err = geno.err)}

  ParOUT_Simp <- sequoia(GenoM = as.matrix(geno.matrix), LifeHistData = lifeHistory.tbl, MaxSibIter = 0,Complex = "simp", Err = geno.err) 
  SeqOUT <- sequoia(GenoM = geno.matrix,
                   SeqList = ParOUT_Simp,
                   MaxSibIter = 20, Err = geno.err, Complex = "simp")
  
  runtime <- toc()
  
  save(ParOUT_Mono, ParOUT_Simp, SeqOUT, file = paste0(outfolder.path, "/sequoia_out.rds",collapse = ""))
  
  run.summary$runtime <- runtime$toc - runtime$tic
  }
   else {
  run.summary$runtime <- pastLS$runtime
  
  load(paste0(outfolder.path, "/sequoia_out.rds",collapse = ""))
   }
  
  id.ls <-readRDS(paste0(outfolder.path, "/id.rds",collapse = ""))$id
  max.id <- max(id.ls)
  
   geno.sim.tbl <-  read.table(paste0(outfolder.path, "/ingeno.txt",collapse = ""), stringsAsFactors = FALSE)
     
    observed.id.ls <- geno.sim.tbl$V1
    
    parentage.grp.truth <- mating.factor.tbl %>% ungroup() %>%
      summarise(kid.id = ifelse(kid %in% observed.id.ls, kid, -1),
             pa.T = ifelse(pa %in% observed.id.ls, pa, -1),
             ma.T = ifelse(ma %in% observed.id.ls, ma, -1),
             id = kid.id) %>%
      filter(kid.id != -1)
    
  if (sf.rate != 0) {
    
    message("preparing parentage assignment tbl ....")
    
  
    
    run.summary$parent.tbl <- left_join(parentage.grp.truth, SeqOUT$PedigreePar %>% as_tibble() %>% select(id, sire, dam, LLRpair) %>% ungroup() %>% mutate(id =as.numeric(id)) , by="id") %>%
      ungroup() %>%
      rename(pa.id = "sire", ma.id = "dam") %>%
      mutate(pa.id = ifelse(is.na(pa.id), -1, as.numeric(pa.id)),
             ma.id = ifelse(is.na(ma.id), -1, as.numeric(ma.id)),
             n.match = (pa.id == pa.T)+(ma.id == ma.T))
  }
  
  
  
  sequoia.ped.1 <- inner_join(SeqOUT$Pedigree %>% filter(id %in% parentage.grp.truth$kid.id),
                              data.frame(id=rownames(geno.matrix))) %>% as_tibble() %>%
    select(id, sire, dam, LLRpair) %>% ## using LLRpair as the metric of ordering
    ungroup() %>%
    mutate(sire = ifelse(is.na(sire), row_number(), sire),
           dam = ifelse(is.na(dam), row_number(), dam)) %>%
    mutate(norm.LLRpair = ifelse(is.na(LLRpair),
                                 0,
                                 exp(LLRpair)),
           prob=norm.LLRpair/max(norm.LLRpair)) %>%
    select(-LLRpair, -norm.LLRpair)
  
   sequoia.ped.2 <- 
     sequoia.ped.1 %>%
     mutate(sire = as.numeric(factor(sequoia.ped.1$sire))+max.id,
           dam = as.numeric(factor(sequoia.ped.1$dam))+2*max.id) %>%
    rename(pa = "sire", ma = "dam", kid = "id")
  
  sib.pairs <- sequoia.ped.2 %>%
    dplyr::group_by(pa, ma) %>%
    dplyr::summarise(full.sib = paste0(kid, collapse = ","),
                     n.sibs = n(),
                     prob = prob[1]) %>%
    dplyr::filter(n.sibs > 1) %>%
    dplyr::ungroup() %>%
    dplyr::select(full.sib, LLRpair) 


  full.sib.grp.infer <- apply(sib.pairs,1,
      function(x) {rbind(combn(as.numeric(unlist(strsplit(x[1],",",fixed=T))),
                              m=2,
                              simplify = T),x[2])}) %>%
    dplyr::bind_cols() %>%
    t %>%
    dplyr::as_tibble() %>% 
    dplyr::mutate(V1 = as.numeric(V1),
                  V2 = as.numeric(V2),
                  V3 = as.numeric(V3)) %>%
    dplyr::mutate(kid.1 = ifelse(V1<V2, V1,V2),
           kid.2 = ifelse(V1<V2, V2,V1)) %>%
    dplyr::group_by(kid.1, kid.2) %>%
    dplyr::summarise(prob=V3[1]) %>%
    dplyr::ungroup() %>%
    dplyr::arrange(desc(prob)) %>%
    dplyr::mutate(kid.1 = as.character(subbingID(as.numeric(kid.1), max.id, id.ls)),
                  kid.2 = as.character(subbingID(as.numeric(kid.2), max.id, id.ls)))

  
  write.table(full.sib.grp.infer,
              paste0(outfolder.path,"/fullsib_seq_assignment.txt"),
              sep = " ",eol = "\n",quote = FALSE, col.names = FALSE, row.names = FALSE)
  
  ## perform pairwise sib-pair from the truth ped
  full.sib.pair.truth.tbl <- RetrieveFullSibTruth(mating.factor.tbl)
  ## run cluster analysis
  
  run.summary$pairwise.tbl <- full_join(full.sib.grp.infer, full.sib.pair.truth.tbl, by= c("kid.1", "kid.2"))
  run.summary$cluster.ls <- RetrieveFullSibGrpSeq(sequoia.ped.2, mating.factor.tbl, max.id, id.ls)
  return(run.summary)
  
}

```



```{r}
RunColony <- function(seed.num = 234,
                        outdir.path="/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib", 
                        replicate.indx  = 1,
                        sf.rate = 1, ped.df = chinook_full_sibs_pedigree,
                      run.name = "colonialTrial",
                      perl.script.path = "/Users/thomasn/repo/pedigree-factor-graphs/src/SimToColony_2065.pl",
                      colony.path = "/Users/thomasn/Downloads/colony2.mac_.20180730_1/colony2s.out",
                      rerun =TRUE,
                      pastLS = NA,
                      case.label=paste0("sf_",sf.rate)) {
  
  if (!rerun && is.na(pastLS) ) {stop("need past LS")}
  
  run.summary <- list()
  
  run.summary$param <- tibble::lst(seed.num, outdir.path, replicate.indx, run.name, sf.rate, perl.script.path)
  
  ## require that pedFac has already run Sim
  ## factorize ped.df
  mating.factor <- factor(unlist(ped.df))
  mating.lvl <- levels(mating.factor)
  mating.factor.tbl <- matrix(as.numeric(mating.factor), ncol=3) %>% as.tibble()
  colnames(mating.factor.tbl) <- c("kid","pa","ma")
  
  outfolder.path <- paste0(outdir.path,"/",case.label,"/",replicate.indx)
  geno.txt.path <- paste0(outfolder.path,"/","geno.txt")
  pedFac.geno.tbl <- read.table(geno.txt.path, stringsAsFactors = F) %>%
    as_tibble()
  
  set.seed(seed.num)
  
  colony.prep.cmd <- paste0("perl ", perl.script.path, " -f ", outfolder.path, " -g 0 -n ",run.name, collapse = "")
  
  if (rerun) system(colony.prep.cmd,ignore.stderr = F, ignore.stdout = F)
  
  if (rerun) dir.create(paste0(outfolder.path,"/colony"), recursive = TRUE)
  if (rerun) file.copy(paste0(outfolder.path,"/colony_0.in"),
            paste0(outfolder.path,"/colony/colony_0.in"), overwrite = T)
            
  colony.run.cmd <- paste0('faketime "2008-01-01 00:00:00" ', colony.path,  " IFN:",outfolder.path,"/colony_0.in >  ",outfolder.path,"/colony/colony_run.out")
  
  if (rerun) {
  tic("run colony:")
  system(colony.run.cmd, ignore.stderr = F, ignore.stdout = F)
  runtime <- toc()
  run.summary$runtime <- runtime$toc - runtime$tic
  } else {
    run.summary$runtime <- pastLS$runtime
  }
  
  run.summary$colony.cmd <- colony.run.cmd 
  
  if (rerun)  colony.run.files <- list.files(".", pattern = paste0(run.name))
  if (rerun)  file.copy(colony.run.files, paste0(outfolder.path, "/colony/."), recursive = T, copy.date=T, overwrite = T) 

  id.ls <-readRDS(paste0(outfolder.path, "/id.rds",collapse = ""))$id
  max.id <- max(id.ls)
  
  
  colony.cluster.tbl <-  read.table(paste0(outfolder.path, "/colony/",run.name,".BestFSFamily",collapse = ""), stringsAsFactors = FALSE, skip = 1) %>% as_tibble()
  colnames(colony.cluster.tbl) <- c("cluster", "prob.incl", "prob.excl", "fullsib")
     

    
  if (sf.rate != 0) {
    message("preparing parentage assignment tbl ....")
    
    colony.infer.tbl <-  read.csv(paste0(outfolder.path, "/colony/",run.name,".ParentPair",collapse = ""), stringsAsFactors = FALSE) %>% as_tibble()
    colnames(colony.infer.tbl) <- c("kid", "pa", "ma", "prob")
  colony.infer.tbl <- colony.infer.tbl %>% mutate(
    pa = ifelse(pa=="*",max.id+1, pa),
    ma = ifelse(ma=="#",max.id+1, ma)
  )
  
 colony.id.infer.tbl <- colony.infer.tbl %>% ungroup() %>%
      mutate(id = subbingID(as.numeric(kid), max.id, id.ls),
             pa.id = subbingID(as.numeric(pa), max.id, id.ls),
             ma.id = subbingID(as.numeric(ma), max.id, id.ls))
    
  
    
    geno.sim.tbl <-  read.table(paste0(outfolder.path, "/ingeno.txt",collapse = ""), stringsAsFactors = FALSE)
    
    observed.id.ls <- geno.sim.tbl$V1
    
    parentage.grp.truth <- mating.factor.tbl %>% ungroup() %>%
      summarise(kid.id = ifelse(kid %in% observed.id.ls, kid, -1),
             pa.T = ifelse(pa %in% observed.id.ls, pa, -1),
             ma.T = ifelse(ma %in% observed.id.ls, ma, -1),
             id = kid.id) %>%
      filter(kid.id != -1)
    
    run.summary$parent.tbl <- left_join(parentage.grp.truth, colony.id.infer.tbl %>% select(id, pa.id, ma.id, prob) %>% ungroup() %>% mutate(id =as.numeric(id)) , by="id") %>%
      ungroup() %>%
      mutate(pa.id = ifelse(is.na(pa.id), -1, as.numeric(pa.id)),
             ma.id = ifelse(is.na(ma.id), -1, as.numeric(ma.id)),
             n.match = (pa.id == pa.T)+(ma.id == ma.T))
  }
  
sib.pairs.ls <- colony.cluster.tbl %>%
    separate_rows(fullsib) %>%
  group_by(cluster) %>%
  filter(n() > 1) %>%
  split(.$cluster) %>%
  map(., 4) %>%
  map(~combn(.x, m = 2)) %>%
  map(~t(.x)) %>%
  map_dfr(as_tibble) %>%
    dplyr::mutate(
      V1=as.numeric(V1),
      V2=as.numeric(V2),
      kid.1 = ifelse(V1<V2, V1,V2),
           kid.2 = ifelse(V1<V2, V2,V1)) %>%
    select(kid.1, kid.2)
  
  full.sib.pair.infer <- left_join(sib.pairs.ls, colony.cluster.tbl %>% separate_rows(fullsib) %>% rename(kid.1 = "fullsib") %>% mutate(kid.1 = as.numeric(kid.1)), by= "kid.1") %>%
    dplyr::mutate(kid.1 = as.character(subbingID(kid.1, max.id, id.ls)),
                  kid.2 = as.character(subbingID(kid.2, max.id, id.ls)))

  
  write.table(full.sib.pair.infer,
              paste0(outfolder.path,"/fullsib_colony_assignment.txt"),
              sep = " ",eol = "\n",quote = FALSE, col.names = FALSE, row.names = FALSE)
  
  ## perform pairwise sib-pair from the truth ped
  full.sib.pair.truth.tbl <- RetrieveFullSibTruth(mating.factor.tbl)
  ## run cluster analysis
  
  run.summary$pairwise.tbl <- full_join(full.sib.pair.infer, full.sib.pair.truth.tbl, by= c("kid.1", "kid.2"))
  run.summary$cluster.ls <- RetrieveFullSibGrpCOLONY(colony.cluster.tbl, mating.factor.tbl, max.id, id.ls)
  return(run.summary)
  
}


```
this block is going to 
```{r, cache=T, eval=FALSE}
sf.array <- c(1, 0.75, 0.5, 0.25, 0)
seed.num <- c(234, 334, 434, 534, 634)

fullsib.pedfac_w_swap <- lapply(1:5, function(sf) {
   lapply(1:5, function (rep) {
    simFulldata(replicate.indx = rep, pedigraph.version = 0.182, pedfac.opt= " -j 1 -f 10", n.iter = 100, seed.num = seed.num[rep], sf.rate = sf.array[sf])
  })
})

#save(fullsib.pedfac_w_swap, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/pedfac_w_swap.Rda")

fullsib.pedfac_w_swap_sub <- lapply(1:5, function(sf) {
   lapply(1:5, function (rep) {
    simFulldata(replicate.indx = rep, pedigraph.version = 0.185, pedfac.opt= " -j 1 -f 10", n.iter = 100, seed.num = seed.num[rep], sf.rate = sf.array[sf], run.label=paste0("swap_sub_sf_",sf,"_",rep))
  })
})

save(fullsib.pedfac_w_swap_sub, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/pedfac_w_swap_sub.Rda")

##rerun some post postprocess 


fullsib.pedfac_w_swap_sub.rerun <- lapply(1:5, function(sf) {
   lapply(1:5, function (rep) {
    simFulldata(replicate.indx = rep, pedigraph.version = 0.185, pedfac.opt= " -j 1 -f 10", n.iter = 100, seed.num = seed.num[rep], sf.rate = sf.array[sf], run.label=paste0("swap_sub_sf_",sf,"_",rep),rerun = FALSE, pastLS = fullsib.pedfac_w_swap_sub[[sf]][[rep]])
  })
})


save(fullsib.pedfac_w_swap_sub.rerun, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/pedfac_w_swap_sub.rerun.Rda")

fullsib.pedfac_5 <- lapply(5, function(sf) {
   lapply(1:5, function (rep) {
    simFulldata(replicate.indx = rep, pedigraph.version = 0.184, pedfac.opt= " -j 1 -f 10", n.iter = 100, seed.num = seed.num[rep], sf.rate = sf.array[sf], run.label=paste0("plain_",sf,"_",rep))
  })
})

save(fullsib.pedfac_5, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/pedfac_5.Rda")


##compare this with Sequoia, COLONY and Franz

fullsib.sequoia <- lapply(1:5, function(sf) {
   lapply(1:5, function (rep) {
    RunSequoia(replicate.indx = rep, seed.num = seed.num[rep], sf.rate = sf.array[sf])
  })
})

save(fullsib.sequoia, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/sequoia_full.Rda")


fullsib.COLONY <- lapply(1:5, function(sf) {
   lapply(1:5, function (rep) {
    RunColony(replicate.indx = rep, seed.num = seed.num[rep], sf.rate = sf.array[sf], run.name=paste0("sf",sf.array[sf],"r",rep))
  })
})

#save(fullsib.COLONY, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/COLONY_full.Rda")


fullsib.sequoia.rerun <- lapply(1:5, function(sf) {
   lapply(1:5, function (rep) {
     message("sf ",sf, "rep ", rep)
    RunSequoia(replicate.indx = rep, seed.num = seed.num[rep], sf.rate = sf.array[sf], rerun=FALSE, pastLS = fullsib.sequoia[[sf]][[rep]])
  })
})

save(fullsib.sequoia.rerun, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/sequoia_full.rerun.Rda")



fullsib.COLONY.rerun <- lapply(1:5, function(sf) {
   lapply(1:5, function (rep) {
    RunColony(replicate.indx = rep, seed.num = seed.num[rep], sf.rate = sf.array[sf], run.name=paste0("sf",sf.array[sf],"r",rep), pastLS = fullsib.COLONY[[sf]][[rep]], rerun = FALSE)
  })
})

save(fullsib.COLONY.rerun, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/COLONY_full.rerun.Rda")



#save(full.sib.pair.truth.tbl, file = "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/full.sib.pair.truth.tbl.Rda")
```

#### setting up for assignment plot

```{r}
load("/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/COLONY_full.Rda")
load("/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/sequoia_full.Rda")
load("/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/pedfac_full.Rda")



sf.array <- c(1, 0.75, 0.5, 0.25, 0)
rep.array <- c(1:5)

X=c(1:4), Y=c(1:5))

###parentage assignment plot ROC (from 1 to 0.25),

all.parentage.tbl <- expand.grid(sf.indx=c(1:4), rep.indx=c(1:5)) %>% as.tibble() %>% mutate(df.result =  map2(.x= sf.indx, .y=rep.indx, .f = function(x,y) {
  pedfac.tem <- fullsib.pedfac[[x]][[y]]$parent.prob.tbl %>% arrange(desc(prob)) %>%
      group_by(kid.id) %>% slice_head(n=1) %>% rename(num.correct.calls = "n.match") %>%
      mutate(pa.id = as.numeric(pa.id), ma.id = as.numeric(ma.id),
             prog.name = "pedFac", n.obs.parents = (pa.T!=-1) +(ma.T!=-1)) %>% ungroup()
  COLONY.tem <- fullsib.COLONY[[x]][[y]]$parent.tbl %>% arrange(desc(prob)) %>%
      group_by(kid.id) %>% slice_head(n=1) %>% rename(num.correct.calls = "n.match") %>%
      mutate(prog.name = "COLONY", n.obs.parents = (pa.T!=-1) +(ma.T!=-1)) %>%
    select(-id) %>% ungroup()
  sequoia.tem <- fullsib.sequoia[[x]][[y]]$parent.tbl %>% ungroup() %>%
      mutate(norm.LLRpair = ifelse(is.na(LLRpair),0, exp(LLRpair)),
             prob=norm.LLRpair/max(norm.LLRpair)) %>%
      arrange(desc(prob)) %>% group_by(kid.id) %>% slice_head(n=1) %>%
      rename(num.correct.calls = "n.match") %>%
      mutate(prog.name = "sequoia", n.obs.parents = (pa.T!=-1) +(ma.T!=-1)) %>%
      select(-id, -norm.LLRpair, -LLRpair) %>% ungroup()
  
  bind_rows(pedfac.tem, COLONY.tem, sequoia.tem)
}))

all.unnested.parentage.tbl <- all.parentage.tbl %>% unnest(cols = df.result)


## creating plots with sf = 1, 0.5, 0 (top - bottom) - of 1 replicate
## sibling cluster plot AUC (column: software)

sf.array <- c(1, 0.75, 0.5, 0.25, 0)

takePairWise <- function(allruns.df, run.label = "") {
map(.x = c(1,3,5), .f = function(x) allruns.df[[x]][[1]]$pairwise.tbl %>%
    ungroup() %>%
    mutate(
      prob = ifelse(is.na(prob), 0, prob),
      is.true = ifelse(is.na(presence.T), 0, 1),
      sf = sf.array[x],
      name = run.label)) %>% bind_rows()
}

pairwise.pack <- bind_rows(
takePairWise(fullsib.pedfac, run.label = "pedFac"),
takePairWise(fullsib.sequoia, run.label = "sequoia"),
takePairWise(fullsib.COLONY, run.label = "COLONY"))



## plot cluster arrangement




```




```{r, eval =FALSE}

  out.ls$errorDistance <- partitionComparison::classificationErrorDistance(new("Partition",
                                                       indiv.cluster.ls$cluster.infer),
                                                   new("Partition",
                                                       indiv.cluster.ls$cluster.truth))

```


```{r, eval =FALSE}
full.sib.roc.input <- lapply(1:5, function(i) {
  roc.input.tbl <- eval(as.name(paste0("fullsib.run",i, collapse="", sep="")))$pairwise.tbl %>%
    ungroup() %>%
    #filter(!is.na(prob)) %>%
    mutate(
      prob = ifelse(is.na(prob), 0, prob),
      is.true = ifelse(is.na(presence.T), 0, 1),
           rep = i) %>%
    rename("posterior" = prob) }) %>% bind_rows()
  
Plot.ROC.sibPair.plot(full.sib.roc.input)
```  

```{r, eval =FALSE}
full.sib.seq.roc.input <- lapply(1:5, function(i) {
  roc.input.tbl <- eval(as.name(paste0("sequoia.run",i, collapse="", sep="")))$pairwise.tbl %>%
    ungroup() %>%
    #filter(!is.na(prob)) %>%
    mutate(
      prob = ifelse(is.na(prob), 0, prob),
      is.true = ifelse(is.na(presence.T), 0, 1),
           rep = i) %>%
    rename("posterior" = prob) }) %>% bind_rows()
  
# normalized LLR - 0 to 1 scale
Plot.ROC.sibPair.plot(full.sib.seq.roc.input %>%
                        ungroup() %>%
                        mutate(posterior = 1-(10^as.numeric(posterior))/max(10^as.numeric(posterior))))
```  

```{r}
```

```{r}
#run1 <- simFulldata(replicate.indx = 2, n.snp = 300)

ll.path <- "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/2/out/ll.txt"
ped.path <- "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib/2/out/ped.txt"

ll.tbl <- read.table(ll.path) %>% as_tibble() %>% 
  group_by(V1) %>%
  mutate(a=row_number())
ll.summary <- ll.tbl %>% group_by(V1) %>% summarise(n.entries = n())
ll.plot.tbl <- left_join(ll.tbl, ll.summary, by="V1") %>%
  mutate(new.index = a/n.entries+V1)
ggplot() + geom_point(data=ll.plot.tbl, aes(x=new.index, y=V3))


lag.ini.tbl <- read.table(ll.path) %>% as_tibble() %>% 
  group_by(V1) %>%
  mutate(a=row_number(), Diff = V3 - lag(V3)) 

ped.mark.tbl <- read.table(ped.path) %>% as_tibble() %>% 
  group_by(V1) %>%
  mutate(a=row_number()) %>%
  group_by(V3, V4) %>%
  mutate(cluster = row_number())

lag.tbl <- left_join(lag.ini.tbl, ped.mark.tbl, by="a")
ggplot() + geom_histogram(data=lag.tbl, aes(x=Diff, fill=cluster==1, position = "fill")) +
  facet_grid(cluster==1~.)

```
When I ran through this, the program was interrupted by an underflow problem with the ongoing message along the edge of the m-node. Eventually, I did rerendition the message propagation algorithm in concieving the carried message value are being carried on in log scale. Much to its dismay, this log transformation carries a cost in performance runtime as this log version of pedFac runs an average of 8-11x slower. An alternative solution is to construct a better prior that encourages pedFac to be more selective in forming half-sib family while the parent generation is not found. 



Testing with the sequoia set
```{r}
out.path <- "/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/halfSib/1"

set.seed(133)
data(Ped_HSg5, LH_HSg5)

err.rate <- 5e-04
error.rate.0 <- 1-(err.rate)-(err.rate/2)^2
error.rate.1 <- (err.rate/2)^2
error.rate.2 <- err.rate /2
error.rate.3 <- 1-err.rate

ErrM <<- matrix(c(error.rate.0,err.rate,error.rate.1,
          error.rate.2, 1-err.rate, error.rate.2,
          error.rate.1,err.rate,error.rate.0), ncol = 3)
        
geno.w.stat.simPedJ2 <- SimGeno(Ped = Ped_HSg5, nSnp = 200, ReturnStats = T)

ParOUT <- sequoia(GenoM = geno.w.stat.simPedJ2$SGeno, LifeHistData = LH_HSg5, MaxSibIter = 0)


SeqOUT <- sequoia(GenoM = geno.w.stat.simPedJ2$SGeno,
                   SeqList = ParOUT,
                   MaxSibIter = 10)

chk <- PedCompare(Ped1 = Ped_HSg5, Ped2 = SeqOUT$Pedigree) 
#chk$Counts["TT",,]

## convert Sequoia's simulated Genotype table into compatible input file for pedFac

id.simPedJ2 <- rownames(geno.w.stat.simPedJ2$SGeno)
n.id.simPedJ2 <- dim(geno.w.stat.simPedJ2$SGeno )[1]
birth.yr.simPedJ2 <- ParOUT$LifeHistPar$BirthYear
geno.w.stat.simPedJ2$SGeno[geno.w.stat.simPedJ2$SGeno==-9] <- 3

afreq.simPedJ2 <- apply(geno.w.stat.simPedJ2$SGeno, 2,
                        function(x) {(sum(x==1)+2*sum(x==2))/(2*sum(x!=3))})

geno.spread.tbl <- geno.w.stat.simPedJ2$SGeno %>%
  as.tibble() %>%
  mutate(id = id.simPedJ2)
  

geno.prefix <- dplyr::bind_cols(id=id.simPedJ2,
                                  is.obs = rep(1, n.id.simPedJ2),
                                  sex = 3-ParOUT$LifeHistPar$Sex,
                                  yr = ParOUT$LifeHistPar$BirthYear,
                                  is.founder = rep(1, n.id.simPedJ2))

geno.join.tbl <- dplyr::left_join(geno.prefix, geno.spread.tbl, by="id")


saveRDS(geno.join.tbl %>% dplyr::select(id) %>% dplyr::ungroup() %>% dplyr::mutate(id.num = dplyr::row_number()),
          paste0(out.path,"/id.rds"))

  geno.join.tbl <- geno.join.tbl %>%
        dplyr::ungroup() %>%
        dplyr::mutate(id = dplyr::row_number())

  write.table(geno.join.tbl %>% arrange(id), paste0(out.path,"/geno.txt"),
              sep = " ",eol = "\n",quote = FALSE, col.names = FALSE, row.names = FALSE)

  write(paste0(c("nIndiv ", n.id.simPedJ2, "\n",
                 "nObsIndiv ", n.id.simPedJ2, "\n",
                 "nGen ",6, "\n",
                 "nSNP ", 200, "\n",
                 "nMar 0\n",
                 "maxID ", n.id.simPedJ2 + 1, "\n",
                 "aFreq ", paste0(round(afreq.simPedJ2,4), collapse = " "), "\n",
                 "epsilon ", paste0(rep(0.001, 200), collapse=" "), "\n",
                 "obsFrac ", paste0(rep(round(n.id.simPedJ2/length(Ped_HSg5$id),3),6), collapse=" "), "\n",
                 "minAge 1\n", 
                 "maxAge 1\n", 
                 "recentYr ", max(birth.yr.simPedJ2), "\n",
                 "maxUnobsLayer ", 1),collapse = ""),
        paste0(out.path,"/prior.txt")
        )

  
  random.arr <- ceiling(c(runif(2,0,1e6)))
  write.table(random.arr, paste0(out.path,"/","rand"),
              sep = " ",eol = " ",quote = FALSE, col.names = FALSE, row.names = FALSE)
  
  pedigraph.path <- "/Users/thomasn/repo/pedigree-factor-graphs/src"
  pedigraph.version <- 0.18
  n.iter <- 10
  pedfac.opt <- ""
  
  pedfac.call <- paste0(pedigraph.path,"/pedigraph_",pedigraph.version, " -d ", out.path,
                         " -n ", n.iter, 
                         " -r ",out.path,"/rand ", pedfac.opt, collapse = "")
  
  
  system(pedfac.call,ignore.stderr = F, ignore.stdout = F)



```


###Verifying sequoia's functionality

Double check whether the manipulation I put on sequoia is error-free. A way to check it is simulate 300 marker with all the parents avail. I presume that it should get most of the assignment right on

```{r}
seed.num = 234
outdir.path="/Users/thomasn/repo/pedigree-factor-graphs/data/ch2/fullsib"
replicate.indx  = 1
sf.rate = 1
ped.df = chinook_full_sibs_pedigree
n.snp =400; geno.err = 0.02; missing.unk = 0.005

seq.check.summary <- list()

  ## factorize ped.df
  mating.factor <- factor(unlist(ped.df))
  mating.lvl <- levels(mating.factor)
  mating.factor.tbl <- matrix(as.numeric(mating.factor), ncol=3) %>% as.tibble()
  colnames(mating.factor.tbl) <- c("kid","pa","ma")
  
  outfolder.path <- paste0(outdir.path,"/sequoia_chk")
  dir.create(file.path(outfolder.path), recursive = TRUE)
  set.seed(234)
  
  geno.input.tbl <- simGeno(mating.tbl = mating.factor.tbl, out.path = outfolder.path,
          n.snp =n.snp, geno.err = geno.err, random.seed = seed.num, sampling.frac.rate = sf.rate, sample.frac.gen = 1, missing.unk = missing.unk) 
  
  pedFac.geno.tbl <- read.table(paste0(outfolder.path,"/","geno.txt"), stringsAsFactors = F) %>%
    as_tibble()
  
  lifeHistory.tbl <- pedFac.geno.tbl %>%
    group_by(V1) %>%
    summarise(ID = V1,
           Sex = as.integer(3-V3),
           BirthYear = V4) %>%
    ungroup() %>%
    select(ID, Sex, BirthYear) %>% data.frame()
  
  geno.matrix <- pedFac.geno.tbl %>% ungroup() %>% select(-V1, -V2, -V3, -V4, -V5) %>%
    as.matrix
  
  geno.matrix[geno.matrix == 3] <- -9
  rownames(geno.matrix) <- lifeHistory.tbl$ID
  
  set.seed(seed.num)
  
  # tic("run sequoia:")
  # 
  # err.rate <- geno.err
  # error.rate.0 <- 1-(err.rate)-(err.rate/2)^2
  # error.rate.1 <- (err.rate/2)^2
  # error.rate.2 <- err.rate /2
  # error.rate.3 <- 1-err.rate

  #ErrM <<- matrix(c(error.rate.0,err.rate,error.rate.1,
   #       error.rate.2, 1-err.rate, error.rate.2,
    #      error.rate.1,err.rate,error.rate.0), ncol = 3)
  
  # ErrM <<- matrix(c(error.rate.3,error.rate.2,error.rate.2,
  #                   error.rate.2, error.rate.3,error.rate.2,
  #                   error.rate.2,error.rate.2,error.rate.3), ncol = 3)
  
  ParOUT_Mono <- sequoia(GenoM = as.matrix(geno.matrix), LifeHistData = lifeHistory.tbl, MaxSibIter = 0,Complex = "mono", Err = geno.err)

  ParOUT_Simp <- sequoia(GenoM = as.matrix(geno.matrix), LifeHistData = lifeHistory.tbl, MaxSibIter = 0,Complex = "simp") 
  SeqOUT <- sequoia(GenoM = geno.matrix,
                   SeqList = ParOUT_Simp,
                   MaxSibIter = 20, Err = geno.err, Complex = "simp")
  
  
  save(ParOUT_Mono, ParOUT_Simp,SeqOUT, file = paste0(outfolder.path, "/sequoia_out.rds",collapse = ""))
  
  id.ls <-readRDS(paste0(outfolder.path, "/id.rds",collapse = ""))$id
  max.id <- max(id.ls)
  
  
    
    message("preparing parentage assignment tbl ....")
    
    geno.sim.tbl <-  read.table(paste0(outfolder.path, "/ingeno.txt",collapse = ""), stringsAsFactors = FALSE)
     
    observed.id.ls <- geno.sim.tbl$V1
    
    parentage.grp.truth <- mating.factor.tbl %>% ungroup() %>%
      summarise(kid.id = ifelse(kid %in% observed.id.ls, kid, -1),
             pa.T = ifelse(pa %in% observed.id.ls, pa, -1),
             ma.T = ifelse(ma %in% observed.id.ls, ma, -1),
             id = kid.id) %>%
      filter(kid.id != -1)
    
  
    left_join(parentage.grp.truth, ParOUT_Mono$PedigreePar %>% as_tibble() %>% select(id, sire, dam, LLRpair) %>% ungroup() %>% mutate(id =as.numeric(id)) , by="id") %>%
      ungroup() %>%
      rename(pa.id = "sire", ma.id = "dam") %>%
      mutate(pa.id = ifelse(is.na(pa.id), -1, as.numeric(pa.id)),
             ma.id = ifelse(is.na(ma.id), -1, as.numeric(ma.id)),
             n.match = (pa.id == pa.T)+(ma.id == ma.T)) %>% count(n.match)
```

```{r}
  
  
  ##not great let me use their simGeno

    sequoia.Ped <- mating.factor.tbl %>%
      rename(id = kid, sire = pa, dam = ma) %>%
      select(id, dam, sire)
    
    sequoia.Ped.compl <- rbind(
    cbind(id=unique(sequoia.Ped$dam), dam=NA, sire=NA),
    cbind(id=unique(sequoia.Ped$sire), dam=NA, sire=NA),
    sequoia.Ped)
    
    
    seq.Geno <- SimGeno(Ped = data.frame(sequoia.Ped), nSnp = 300, SnpError=0.02, ParMis= 0)
      
    ParOUT_Mono <- sequoia(GenoM = seq.Geno, LifeHistData = lifeHistory.tbl, MaxSibIter = 0,Complex = "mono")
    
    left_join(parentage.grp.truth, ParOUT_Mono$PedigreePar %>% as_tibble() %>% select(id, sire, dam, LLRpair) %>% ungroup() %>% mutate(id =as.numeric(id)) , by="id") %>%
      ungroup() %>%
      rename(pa.id = "sire", ma.id = "dam") %>%
      mutate(pa.id = ifelse(is.na(pa.id), -1, as.numeric(pa.id)),
             ma.id = ifelse(is.na(ma.id), -1, as.numeric(ma.id)),
             n.match = (pa.id == pa.T)+(ma.id == ma.T)) %>% count(n.match)
    
```

# A tibble: 3 x 2
  n.match     n
    <int> <int>
1       0   230
2       1   647
3       2   280

```{r}
ParOUT_SIMPL <- sequoia(GenoM = seq.Geno, LifeHistData = lifeHistory.tbl, MaxSibIter = 10,Complex = "simp")
    
    left_join(parentage.grp.truth, ParOUT_SIMPL$PedigreePar %>% as_tibble() %>% select(id, sire, dam, LLRpair) %>% ungroup() %>% mutate(id =as.numeric(id)) , by="id") %>%
      ungroup() %>%
      rename(pa.id = "sire", ma.id = "dam") %>%
      mutate(pa.id = ifelse(is.na(pa.id), -1, as.numeric(pa.id)),
             ma.id = ifelse(is.na(ma.id), -1, as.numeric(ma.id)),
             n.match = (pa.id == pa.T)+(ma.id == ma.T)) %>% count(n.match)
    
```


```{r}
seq.Geno <- SimGeno(Ped = data.frame(sequoia.Ped), nSnp = 200, SnpError=0.02, ParMis= 0)
      
ErrM <<- matrix(c(0,0,0,
                    0, 0,0,
                    0,0,0), ncol = 3)

    ParOUT_Mono <- sequoia(GenoM = seq.Geno, LifeHistData = lifeHistory.tbl, MaxSibIter = 0,Complex = "mono", Err= 0.02)
    
    left_join(parentage.grp.truth, ParOUT_Mono$PedigreePar %>% as_tibble() %>% select(id, sire, dam, LLRpair) %>% ungroup() %>% mutate(id =as.numeric(id)) , by="id") %>%
      ungroup() %>%
      rename(pa.id = "sire", ma.id = "dam") %>%
      mutate(pa.id = ifelse(is.na(pa.id), -1, as.numeric(pa.id)),
             ma.id = ifelse(is.na(ma.id), -1, as.numeric(ma.id)),
             n.match = (pa.id == pa.T)+(ma.id == ma.T)) %>% count(n.match)

```
